### PR TITLE
Implement 2-cohomology and module computations for arbitrary finite groups, not just solvable ones, via `TwoCohomologyGeneric`

### PR DESCRIPTION
**Pull request by hulpke**
Uses rewriting system since pc presentation will not always be available.
This is both code for the cohomology group, and for constructing the extension corresponding to a particular cocycle. Group is constructed by default as Fp group, but there is an option to have it compute a faithful permutation representation.

Also includes improved routines for `IrreducibleModules` for the non-pc case once such cohomology routines exist, and action of compatible pairs on the cohomology group.

There is a test in `testextra` that, as application, constructs perfect groups anew. (This could easily go to larger orders, if desired, at the moment two quick tests are in.)

### CODE REVIEW

File: doc/ref/grppc.xml [hunk 1]
```diff
@@ -417,6 +417,8 @@ The only method with is special for pc groups is a method
 to compute intersections of subgroups, since here a pcgs of a parent
 group is needed and this can only by guaranteed within pc groups.
 
+Section <Ref Sect="2-Cohomology"> describes operations and methods for
```
Comment (fingolfin) [inline doc/ref/grppc.xml]:
```suggestion
Section <Ref Sect="2-Cohomology"/> describes operations and methods for
```

Comment (hulpke) [reply doc/ref/grppc.xml]: Thanks.

**Comment by wilfwilson** [comment]
@hulpke I noticed you’ve removed the WIP from the title and the ‘do not review’ label. Just to check, for sure: does this mean that you’d be happy for the PR to be reviewed, and hopefully merged? If so, I’d suggest removing the ‘do not merge’ label because I get the impression it might put people off from reviewing. Sorry I’ve not had chance to look at your PR further yet.

**Comment by alex-konovalov** [comment]
Travis reports two manual example diffs:
```
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
992)
# Input is:
g1:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
999)
# Input is:
g2:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
```

**Comment by hulpke** [comment]
@wilfwilson 
At this stage I am happy for comments about potential problems, but there is still some code (basic functionality for compatible pairs and action on the cohomology group) I am still working on and want to add before merging. (There also is a manual bug and a stupid typo in the perfect group construction example that need fixing.)
As there will still be code added (as soon as I fix a bug..), it might be best to wait with formal reviewing.

**Comment by hulpke** [comment]
@wilfwilson OK, now the promised changes are in, and the holds are removed.
Since there is no need to backport, I'm not sure whether there is a need to squash into one commit, but could do so if desired.

**Review by fingolfin — CHANGES_REQUESTED** [review]
Very nice, thank you! Overall, this is of course a very welcome improvement!

Unfortunately I did not yet have time to look at everything (with easter and everything), but I left a few rather minor remarks. I would like to have a closer look at a few more things, I'll try to get this done this week or early next week.

Of course it would be nice to have this in GAP 4.11 -- @alex-konovalov is in charge of that.

File: lib/grpreps.gi [hunk 2]
```diff
@@ -44,21 +44,194 @@ end);
...
+
+  mats:=modu.generators;
+  bas:=One(modu.generators[1]);
+  rad:=0;
+
+  if Size(si)>1 then
+
+    # get augmentation ideal of subgroup, at least approximate from
+    # generators
+    a:=AsList(G);
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/grpreps.gi]: Will change so (but it anyhow is a fallback method, permutation groups use Burnside/Brauer method below).

File: lib/gpprmsya.gi [hunk 3]
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
...
+# A_n = < x_1, ..., x_{n-2} | x_1^3, x_i^2 (i > 1),
+#                           (x_i x_{i+1})^3 (1 <= i <= n-3)
+#                           (x_i x_j)^2  (1 <= i, i+1 < j, j <= n-2) >.
+InstallMethod(IsomorphismFpGroupForRewriting,"alternating",
+  [IsNaturalAlternatingGroup],0,
+function ( G )
+local   F,      # free group
+	gens,	#generators of F
+	imgs,
+	hom,	# bijection
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpprmsya.gi]: Done.

File: lib/gpprmsya.gi [hunk 4]
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
...
+	mov,deg,
+	relators,
+	i, j;       # loop variables
+
+    mov:=MovedPoints(G);
+    deg:=Length(mov);
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpprmsya.gi]: Changed

File: lib/gpprmsya.gi [hunk 5]
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
...
+
+    # compute the bijection
+    imgs:=[];
+    Add(imgs,(mov[1],mov[2],mov[3]));
+    for i in [3..deg-1] do
+      Add(imgs,(mov[1],mov[2])(mov[i],mov[i+1]));
+    od;
+
+    # return the isomorphism to the finitely presented group
+    hom:= GroupHomomorphismByImages(G,F,imgs,GeneratorsOfGroup(F));
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpprmsya.gi]: Done

File: lib/gpprmsya.gi (hunk 3 above)
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpprmsya.gi]: I have replaced tabs.

File: lib/gpprmsya.gi (hunk 5 above)
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpprmsya.gi]: No, `gens` are free generators, while `F` now is fp.

File: lib/twocohom.gi [hunk 6]
```diff
@@ -691,7 +853,621 @@ function( G, M )
...
+  for r in tzrules do
+    if Length(r[1])>=2 then
+      Add(rules,r);
+      if Length(r[1])>2 or
+        (Length(r[1])=2 and (Length(r[2])>0 or formalinverse[r[1][1]]<>r[1][2]))
+        then 
+          AddSet(hastail,Length(rules));
+      fi;
+    elif Length(r[1])>1 then
+      if Length(r[2])=0 then Error("generator is trivial");fi;
```
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi [hunk 7]
```diff
@@ -691,7 +853,621 @@ function( G, M )
...
+
+  rulpos:=List(tzrules,x->PositionProperty(rules,y->y[1]=x[1]));
+
+  htpos:=List([1..Length(rules)],x->Position(hastail,x));
+
+  model:=ValueOption("model");
+  if model<>fail then
+    q:=GQuotients(model,G)[1];
+    pre:=List(gens,x->PreImagesRepresentative(q,x));
+    ker:=KernelOfMultiplicativeGeneralMapping(q);
```
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi [hunk 8]
```diff
@@ -691,7 +853,621 @@ function( G, M )
...
+    # rule that would get the tail
+    Add(new,LeftQuotient(mapped2(rules[i][1]),mapped2(rules[i][2])));
+  od;
+  r.presentation:=rec(group:=FreeGroupOfFpGroup(fpg),relators:=new,
+    prewords:=List(ogens,x->UnderlyingElement(ImagesRepresentative(fp,x))));
+
+  r.pairact:=function(zy,pair)
+  local autom,mat,i,imagemonwords,imgwrd,left,right,extim,prdout,myinvers,v;
+
+    autom:=InverseGeneralMapping(pair[1]);
```
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi [hunk 9]
```diff
@@ -691,7 +853,621 @@ function( G, M )
...
+    new:=new*SmallerDegreePermutationRepresentation(p);
+    SetIsomorphismPermGroup(fp,new);
+  fi;
+
+  return fp;
+end);
+
+InstallGlobalFunction("CompatiblePairOrbitRepsGeneric",function(cp,coh)
+local bas,ran,mats,o;
+  if Length(coh.cohomology)=0 then return [coh.zero];fi;
```
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/grppcext.gi [hunk 10]
```diff
@@ -497,13 +497,58 @@ end);
...
     Mgrp := GroupByGenerators( M.generators );
-    oper := GroupHomomorphismByImagesNC( G, Mgrp, Pcgs(G), M.generators );
+    Ggens:=Pcgs(G);
+    oper:=fail;
+    if IsPcgs(Ggens) and Length(Ggens)=Length(M.generators) then
+      oper := GroupHomomorphismByImagesNC( G, Mgrp, Ggens, M.generators );
+    elif Length(arg)=2 then
+      # search through automorphism group for projection image and reps,
+      # then add module automorphisms
+      gens:=GeneratorsOfGroup(G);
```
[the remark this answers is not in the document]
Comment (alex-konovalov) [reply lib/grppcext.gi]: I could run the test of this PR in Jenkins, to exercise testextra. However, it does not pass cleanly at the moment: #3408

[the remark this answers is not in the document]
Comment (hulpke) [reply lib/grppcext.gi]: @fingolfin I have added smaller a test to `testinstall' that checks for action of compatible pairs (thus testing most code). Is seems to run through (the current test failure is due to a DNS resolving problem).

[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/grppcext.gi]: Thanks. I'll go over the coverage once again.

File: lib/grppcext.gi [hunk 11]
```diff
@@ -497,13 +497,58 @@ end);
...
+          Add(pool,DirectProductElement([aut,mat]));
+          return true;
+        fi;
+        return false;
+      end;
+      K:=SubgroupProperty(Image(triso),test);
+      B:=MTX.ModuleAutomorphisms(M);
+      if Size(B)>1 then
+        for i in Set(GeneratorsOfGroup(B)) do
+          Add(pool,DirectProductElement([One(A),i]));
```
Comment (fingolfin) [inline lib/grppcext.gi]: This loop is still not covered.

File: lib/grppcext.gi [hunk 12]
```diff
@@ -497,13 +497,58 @@ end);
...
+      else
+        A:=GroupWithGenerators(pool);
+        SetSize(A,Size(K)*Size(B));
+      fi;
+      return A;
+    fi;
+
+    if oper=fail then
+      Ggens:=GeneratorsOfGroup(G);
+      oper := GroupHomomorphismByImagesNC( G, Mgrp, Ggens, M.generators );
```
Comment (fingolfin) [inline lib/grppcext.gi]: Still not covered.

File: lib/grppcext.gi [hunk 13]
```diff
@@ -635,11 +681,27 @@ local G, M, Mgrp, oper, A, B, D, translate, gens, genimgs, triso, K, K1,
...
     end;
 
-    if Size(G)>20000 then
-      # if G is too large we cannot write out elements
-      elmlist:=fail; 
-      f:=basicact;
-    else
+    if not IsPcgs(Ggens) then
+      elmlist:=fail;
+      epi:=EpimorphismFromFreeGroup(G);
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/grppcext.gi]: Still not covered.

File: lib/grppcext.gi [hunk 14]
```diff
@@ -648,151 +710,157 @@ local G, M, Mgrp, oper, A, B, D, translate, gens, genimgs, triso, K, K1,
...
+            gens:=translate.permgens;
+            genimgs:=translate.pairgens;
+            triso:=translate.isomorphism;
+            D:=translate.permgroup;
+          fi;
+        fi;
+      until l=1;
+    else
+      #D:=Stabilizer(D,M.generators,gens,genimgs,f);
+      hf:=SparseIntKeyVecListAndMatrix(false,Concatenation(M.generators));
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/grppcext.gi]: Still not covered.

File: lib/grpreps.gi [hunk 15]
```diff
@@ -44,21 +44,195 @@ end);
...
+  if dim=1 then
+    # linear representations come from G/G'
+    gens:=GeneratorsOfGroup(G);
+    a:=DerivedSubgroup(G);
+    if Size(a)=Size(G) then 
+      return [gens,[TrivialModule(Length(gens),F)]];
+    else
+      a:=MaximalAbelianQuotient(G);
+      si:=List(gens,x->ImagesRepresentative(a,x));
+      sub:=IrreducibleModules(Group(si),F,1);
```
Comment (fingolfin) [inline lib/grpreps.gi]: Still not covered.

File: lib/grpreps.gi [hunk 16]
```diff
@@ -44,21 +44,195 @@ end);
...
+
+  fi;
+
+  modu := RegularModule( G, F );
+  gens:=modu[1];
+  modu:=modu[2];
+
+  # The augmentation ideal of a P-normal subgroup lies in the radical
+  if Characteristic(F)=0 then
+    si:=TrivialSubgroup(G);
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/grpreps.gi]: Still not covered.

File: lib/grpreps.gi [hunk 17]
```diff
@@ -44,21 +44,195 @@ end);
...
+  od;
+
+  Info(InfoMeatAxe,1,"submodules:",List(subs,Length));
+
+  if Length(subs)>0 then
+    # close under sums/intersections to form series
+    ser:=[[],subs[1],One(mats[1])];
+    for i in [2..Length(subs)] do
+      a:=subs[i];
+      j:=2;
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/grpreps.gi]: Still not covered.

File: lib/twocohom.gi [hunk 18]
```diff
@@ -667,6 +667,168 @@ function( G, M )
...
+      ffp:=Range(dec.homs[i]);
+    fi;
+    imo:=IsomorphismFpMonoid(ffp);
+    Add(mos,imo);
+    m:=Range(imo);
+    loff:=off-Length(GeneratorsOfMonoid(m));
+    monreps:=fmgens{[loff+1..off]};
+    monreal:=mgens{[loff+1..off]};
+    if IsBound(m!.rewritingSystem) then
+      k:=m!.rewritingSystem;
```
Comment (fingolfin) [inline lib/twocohom.gi]: This is one of only two lines in this function which is no covered. Very nice!

File: lib/twocohom.gi [hunk 19]
```diff
@@ -667,6 +667,168 @@ function( G, M )
...
+        end,
+        function(w)
+          local g,i,x;
+          g:=[];
+          for i in LetterRepAssocWord(UnderlyingElement(w)) do
+            if IsOddInt(i) then x:=(i+1)/2;
+            else x:=-i/2;fi;
+            # word must be freely cancelled
+            if Length(g)>0 and x=-g[Length(g)] then
+              Unbind(g[Length(g)]);
```
Comment (fingolfin) [inline lib/twocohom.gi]: (This is the other uncovered line).

File: lib/twocohom.gi [hunk 20]
```diff
@@ -691,7 +853,621 @@ function( G, M )
...
+      c,nvars,htpos,zeroq,r,ogens,bds,model,q,pre,pcgs,miso,ker,solvec,rulpos;
+
+
+  # collect the word in factor group
+  colltz:=function(a)
+  local i,j,s,mm,p;
+
+    # collect from left
+    i:=1;
+    while i<=Length(a) do
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply lib/twocohom.gi]: This is now covered, thanks.

File: lib/twocohom.gi (hunk 6 above)
Comment (fingolfin) [reply lib/twocohom.gi]: Still not covered.

File: lib/twocohom.gi (hunk 8 above)
Comment (fingolfin) [reply lib/twocohom.gi]: This is now covered, nice!

**Review by fingolfin — APPROVED** [review]
Thank you for the added `testinstall` test. There are still about 300 new lines of code which are not covered by a test in `testinstall`, and the new code only has a coverage rate of ~76% as a result. I'd prefer if it was higher (not because high coverage by itself is a magic goal, but rather because it would increase my confidence in all those special cases, and also in our ability to keep this code from breaking in the future).

Anyway, I don't think we need to delay this further because of that. So let's merge it soon. (I'll wait for Travis to finish with its current truck loads of builds before merging it).

File: lib/twocohom.gi (hunk 18 above)
Comment (hulpke) [reply lib/twocohom.gi]: Yes, this is as a hook for future extensions (RWS for simple groups).

**Pull request merged**